### PR TITLE
Pluto notebook with GSLIB comparison

### PR DESCRIPTION
**Pull request by exepulveda**
Finally the GSLIB comparison. I don't know why markdown is not properly displaying URL.

### CODE REVIEW

**Review by juliohm — CHANGES_REQUESTED** [review]
Thank you @exepulveda , can you please use the same conventions of the other notebooks? Notice that all notebooks start with a few setup cells that are hidden. Please copy these cells to your notebook as well.

Also, please rename GSLIB_Comparison.jl to GSLIBComparison.jl to follow the file name conventions. Notice that you have written a single big cell with a lot of markdown in it. I would split this cells into multiple cells with one section per cell to facilitate editing.

Currently the notebook is not compiling because the configuration cells I mentioned aren't present.

**Comment by exepulveda** [comment]
Is there any readme.md with the conversions for next time?
Why is the nootebook runnig but you say it does not compile?

I will make the changes that you asked for.

**Comment by juliohm** [comment]
> Is there any readme.md with the conversions for next time?

Is it too hard to figure out that none of the notebooks in this repository has a name with underscores in it? Or that all notebooks have a common header setting up the environment? I thought that these details were easy to catch. To be honest, I feel that you are submitting PRs without the necessary attention. As a result, we both waste a lot of time fixing formatting issues.

I can spend time writing a "convention" file, but I doubt it will be more useful than just copying/pasting the conventions from existing notebooks.

> Why is the nootebook runnig but you say it does not compile?

You are still having trouble with basic environment settings. This repository has a `Project.toml` and a `Manifest.toml` in the `notebooks` folder. This means that the `notebooks` folder is a Julia project. Every notebook in this folder has a set of cells in the beginning activating the environment of the project. When I run your notebook it doesn't activate the environment so I can't even import GeoStats.jl

> I will make the changes that you asked for.

Thanks! Try to be more careful next time.

**Review by juliohm — CHANGES_REQUESTED** [review]
I've noticed that you have *.gslib and *.csv files in this PR. Can you please confirm that they are different data sets before I spend time trying to figure out the differences?

In case they are the same data set saved in two different formats, I hope you understand we don't need both.

**Comment by exepulveda** [comment]
> I've noticed that you have *.gslib and *.csv files in this PR. Can you please confirm that they are different data sets before I spend time trying to figure out the differences?
> 
> In case they are the same data set saved in two different formats, I hope you understand we don't need both.

They are the same, but different formats, because users may want to reproduce the GSLIB results, therefore I'm providing with all necessary to do so. This notebook is about that. Of course if you want not to use the csv as the standard used in all tutorials in this project, we would need to load the glib file instead, which in my opinion is not desirable.

**Comment by juliohm** [comment]
We can load the gslib file with GslibIO.jl in theory and that is showing another important feature of the stack. Users coming from gsblib can just load the file without converting to csv. Can we rewrite the example using GslibIO.jl?

**Comment by exepulveda** [comment]
> > Is there any readme.md with the conversions for next time?
> 
> Is it too hard to figure out that none of the notebooks in this repository has a name with underscores in it? Or that all notebooks have a common header setting up the environment? I thought that these details were easy to catch. To be honest, I feel that you are submitting PRs without the necessary attention. As a result, we both waste a lot of time fixing formatting issues.
> 
> I can spend time writing a "convention" file, but I doubt it will be more useful than just copying/pasting the conventions from existing notebooks.
> 
> > Why is the nootebook runnig but you say it does not compile?
> 
> You are still having trouble with basic environment settings. This repository has a `Project.toml` and a `Manifest.toml` in the `notebooks` folder. This means that the `notebooks` folder is a Julia project. Every notebook in this folder has a set of cells in the beginning activating the environment of the project. When I run your notebook it doesn't activate the environment so I can't even import GeoStats.jl
> 
> > I will make the changes that you asked for.
> 
> Thanks! Try to be more careful next time.

I just asked two simple questions, because I think is good to have a readme.md (or simular) with used conventions for developers as many other projects have. Sometimes it is very hard to deduce thinks.

Every one needs to be more careful and be kind, if not, there is not a good environment for collaboration.

In the last push, I have fixed all you indicated. Apparently, the only pending is the data with both csv and glib formats.

**Comment by juliohm** [comment]
> Every one needs to be more careful and be kind, if not, there is not a good environment for collaboration.

That is for sure. Being professional is a must, especially in the open internet. We need to point out inefficiencies when we see them, otherwise they will never go away. The ultimate goal is to succeed together.

We will likely need to iterate many more times in this PR to convey a clear message and to highlight the features of the project.

**Comment by exepulveda** [comment]
> We will likely need to iterate many more times in this PR to convey a clear message and to highlight the features of the project.

Happy to take into account your suggestions.

**Comment by exepulveda** [comment]
> We can load the gslib file with GslibIO.jl in theory and that is showing another important feature of the stack. Users coming from gsblib can just load the file without converting to csv. Can we rewrite the example using GslibIO.jl?

I need to check if GslibIO.jl is working properly, because it does not load indeed regular grids generated by kt3d without fixing the gslib file a little bit. I will report this problem very soon.

**Comment by juliohm** [comment]
That is a good outcome. If `GslibIO.load_legacy` is not working with all files, we need to fix it first.

**Comment by exepulveda** [comment]
The problem with load_legacy is that it only loads regular grid, but not generic GSLIB files (x, y, z, var1, var2,... etc).
Also, I noticed that you extended?changed the Gslib format to include grid specifications and more, but it is misleading. It would be named load_extended or something and load_legacy just load (which references to the actual GSLIB format).

Any GSLIB user would expect to work with the standar GSLIB format not the extended, which is not compatible with gslib programs.

**Comment by juliohm** [comment]
I think we need to address two issues:

1. Add support for generic GSLIB in `load_legacy`. If you know the specification, could you please submit a PR? This PR could also include a new `save_legacy` to save in the old format.

2. Document `load_legacy` and `save_legacy` better in the README so that newcomers can easily interop with GSLIB without changing the file format they are already familiar with. I don't think we want to make this legacy format the default given that it is incomplete and requires user input. Users may stick to `load_legacy/save_legacy` if they are planning to keep GSLIB in the loop, or `load/save` if all they want is move on to a more self-contained ASCII format. I myself use a combination of `load_legacy/save` to load old files and rewrite to disk in a self-contained format.

**Comment by exepulveda** [comment]
I have submitted a PR with legacy load/save functions.
For large block models, storing models in text file is discouraged. Just imagine a simulation for a block model of hundreds of millions, that can easily end with a gigantic text file. I more inclined to use HDF5 files (or perhaps Apache Arrow), but that requires a lot of volunteers' time and dedication. 

Could you review the PR in the GslibIO project? If we can merge this, I can apply it to this Pluto notebook.

**Comment by juliohm** [comment]
Yes. I fully agree. Will review the PR there by end of week or beginning of
next week. I'm far from my computer this week.

Thanks for the improvements.

File: notebooks/GSLIBComparison.jl [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+  using Plots; gr(size=(700,400))
+end
+
+
+# ╔═╡ 524c65d0-2ea9-11eb-0450-5ff97535406a
+md"""
+# GSLIB comparison
+
+This tutorial shows how to run a Kriging estimation problem that homologates GSLIB kriging program *kt3d*.
+
```
Comment (rmcaixeta) [inline notebooks/GSLIBComparison.jl:42]: Wouldn't it be an homologation of GeoStats.jl using kt3d as reference?

**Review by rmcaixeta — COMMENTED** [review]
Nice results @exepulveda. Just made a note about homologation for eventual discussion
Curiosity: GSLIB took around 10s for kriging; did you benchmark GeoStats.jl too?

File: notebooks/GSLIBComparison.jl (hunk 1 above)
Comment (juliohm) [reply notebooks/GSLIBComparison.jl:42]: I agree @rmcaixeta , I think the English word `homologation` applies to GeoStats.jl with GSLIB as a reference. The sentence written above seems to state the opposite direction.

Comment (exepulveda) [reply notebooks/GSLIBComparison.jl:42]: Maybe it is better to say "reproduce" instead "homologate". Or perhaps a comparison only.

**Comment by exepulveda** [comment]
> Nice results @exepulveda. Just made a note about homologation for eventual discussion
> Curiosity: GSLIB took around 10s for kriging; did you benchmark GeoStats.jl too?

It is faster. It should be because GSLIB does not use the best algorithms (spatial indexing and matrix factorisation). I should report the elapsed time of GeoStats too.

File: notebooks/GSLIBComparison.jl (hunk 1 above)
Comment (juliohm) [reply notebooks/GSLIBComparison.jl:42]: Reproduce is a simpler and better term :100:

**Comment by juliohm** [comment]
After https://github.com/JuliaEarth/GslibIO.jl/pull/10 is ready, we can come back and polish this PR.

There are at least 3 important messages we want to convey with this notebook:

1. We can easily load/save GSLIB files in a workflow
2. GeoStats.jl is faster and more flexible than GSLIB
3. The results obtained with both projects match

Of course these sentences are bold, and we just need to use the correct words in the notebook. For example, *more flexible* here means that users can easily interact with different options without having to open a text file to set parameters manually. And *faster* here means that the 3D kriging solver is faster with this dataset and setup. You get the point...

**Comment by exepulveda** [comment]
I would like to also highlight that there is still a missing functionality in GeoStats that makes it not yet 100% compatible with kt3d: block kriging. We need to say how we can overcome this until GeoStats can offer this functionality.

**Comment by juliohm** [comment]
Fully agree. I am actively working in Meshes.jl to generalize the domain representations, but it will take some time until we get there. It will be there in 2021 for sure 👍🏽

**Comment by juliohm** [comment]
@exepulveda do you plan to finish this PR?

**Comment by exepulveda** [comment]
> @exepulveda do you plan to finish this PR?

Yes, I do. I need to finish some work first. I guess in two weeks I would have the time.

**Comment by juliohm** [comment]
Closing the PR as stale after the proposed waiting period of two weeks.

**Pull request closed without merging**